src/GraphQL.MicrosoftDI/GraphQLBuilderExtensions.cs

```diff
@@ -0,0 +1,72 @@
+        /// implementation does not support subscriptions.
+        /// <br/><br/>
+        /// Also configures <see cref="ComplexityConfiguration"/> to be pulled from the dependency
+        /// injection provider, overwriting values within <see cref="ExecutionOptions.ComplexityConfiguration"/>
+        /// with values configured within the registered instance if set there.
+        /// </summary>
+        public static IGraphQLBuilder AddGraphQL(this IServiceCollection services)
+            => new GraphQLBuilder(services);
+
+        /// <summary>
+        /// Registers <typeparamref name="TSchema"/> within the dependency injection framework. <see cref="ISchema"/> is also
+        /// registered if it is not already registered within the dependency injection framework. Services required by
+        /// <typeparamref name="TSchema"/> are instianted directly if not registered within the dependency injection framework.
+        /// This can eliminate the need to register each of the graph types with the dependency injection framework, either
+        /// manually or via <see cref="GraphQL.GraphQLBuilderExtensions.AddGraphTypes(IGraphQLBuilder)"/>. Singleton and scoped
+        /// lifetimes are supported.
+        /// </summary>
+        /// <remarks>
+        /// Schemas that implement <see cref="IDisposable"/> of a transient lifetime are not supported, as this will cause a
+        /// memory leak if requested from the root service provider.
+        /// </remarks>
+        public static IGraphQLBuilder AddSelfActivatingSchema<TSchema>(this IGraphQLBuilder builder, ServiceLifetime serviceLifetime = ServiceLifetime.Singleton)
+            where TSchema : class, ISchema
+        {
+            if (serviceLifetime == ServiceLifetime.Transient && typeof(IDisposable).IsAssignableFrom(typeof(TSchema)))
+            {
+                // This scenario can cause a memory leak if the schema is requested from the root service provider.
+                // If it was requested from a scoped provider, then there is no reason to register it as transient.
+                // See following link:
+                // https://docs.microsoft.com/en-us/dotnet/core/extensions/dependency-injection-guidelines#disposable-transient-services-captured-by-container
+                throw new InvalidOperationException("A schema that implements IDisposable cannot be registered as a transient service.");
+            }
+
+            // Register the service with the DI provider as TSchema, overwriting any existing registration
+            builder.Register(services =>
+            {
+                var selfActivatingServices = new SelfActivatingServiceProvider(services);
+                var schema = ActivatorUtilities.CreateInstance<TSchema>(selfActivatingServices);
+                return schema;
+            }, serviceLifetime);
+
+            // Now register the service as ISchema if not already registered.
+            builder.TryRegister<ISchema>(services =>
```

Comment:
```csharp
 builder.TryRegister<ISchema>(services =>
            {
                var selfActivatingServices = new SelfActivatingServiceProvider(services);
                var schema = ActivatorUtilities.CreateInstance<TSchema>(selfActivatingServices);
                return schema;
            }, serviceLifetime);

           return builder;
```
->
```c#
return builder.TryRegister<ISchema>(services => services.GetRequiredService<TSchema>(), serviceLifetime);
```

Reply:
I thought about this and decided against it, since it will be registering a service for disposal that it actually doesn't own, if you understand what I mean.  Perhaps since the lifetime is the same here between ISchema and TSchema then it would not matter.  I was not sure so I figured it would be safer to simply duplicate the couple lines of code over again.

I think perhaps there was other areas where I did use the pattern you suggested - I can't remember now.  I know I thought a while about it.  But here I felt it was better not to do so.

This is an area where Autofac can do better than MS DI -- with Autofac you can register a single factory/instance as multiple service types.  (Not that it helps us here, just FYI.)

I'm open to comments here.

Reply:
Note: Part of the decision was also based on the fact that the only time TSchema or ISchema is typically pulled from DI is during execution.  It's not like we have other classes that pull ISchema and might be pulling a different instance than the execution which might pull TSchema.  At least I couldn't think of any scenario.

Reply:
> I thought about this and decided against it, since it will be registering a service for disposal that it actually doesn't own, if you understand what I mean.

I understand. This is not the first place where I see this. I came across such doubled registration in other projects. As a rule, this happened in the result of searching for bugs. We need to choose the smallest evil. I am sure that there will be more problems from double registration and double creating an object (singleton!).

Reply:
>  Perhaps since the lifetime is the same here between ISchema and TSchema then it would not matter.

Sure.

Reply:
#2797